aloha-core/src/main/scala/com/eharmony/aloha/models/multilabel/MultilabelModel.scala

```diff
@@ -0,0 +1,387 @@
+    * @tparam U upper bound on model output type `B`
+    * @tparam K type of label or class
+    * @tparam B output type of the model.
+    * @return a SubValue indicating success.
+    */
+  protected[multilabel] def reportSuccess[U, K, B <: U](
+      modelId: ModelIdentity,
+      labelInfo: LabelsAndInfo[K],
+      missing: scm.Map[String, Seq[String]],
+      prediction: Map[K, Double],
+      auditor: Auditor[U, Map[K, Double], B]
+  ): Subvalue[B, Map[K, Double]] = {
+
+    val aud = auditor.success(
+      modelId,
+      prediction,
+      errorMsgs = labelInfo.errorMsgs,
+      missingVarNames = combineMissing(labelInfo, missing)
+    )
+
+    Subvalue(aud, Option(prediction))
+  }
+
+  /**
+    * Report that a `Throwable` was thrown while invoking the predictor
+    * @param modelId An identifier for the model.  Used in error reporting.
+    * @param labelInfo labels and information about the labels.
+    * @param missing missing features from
+    * @param throwable the error the occurred in the predictor.
+    * @param auditor an auditor used to audit the output.
+    * @tparam U upper bound on model output type `B`
+    * @tparam K type of label or class
+    * @tparam B output type of the model.
+    * @return a SubValue indicating failure.
+    */
+  protected[multilabel] def reportPredictorError[U, K, B <: U](
+      modelId: ModelIdentity,
+      labelInfo: LabelsAndInfo[K],
+      missing: scm.Map[String, Seq[String]],
+      throwable: Throwable,
+      auditor: Auditor[U, Map[K, Double], B]
+  ): Subvalue[B, Nothing] = {
+
+    val pw = new PrintWriter(new StringWriter)
+    throwable.printStackTrace(pw)
+    val stackTrace = pw.toString.split("\n").take(NumLinesToKeepInStackTrace).mkString("\n")
+
+    val aud = auditor.failure(
+      modelId,
+      errorMsgs = stackTrace +: labelInfo.errorMsgs,
+      missingVarNames = combineMissing(labelInfo, missing)
+    )
+    Subvalue(aud, None)
+  }
+
+  /**
+    * Get labels from the input for which a prediction should be produced.
+    * @param example the example provided to the model
+    * @param labelsOfInterest a function used to extract labels for which a
+    *                         prediction should be produced.
+    * @param labelToInd mapping from Label to index into the sequence of all
+    *                   labels seen in the training set.
+    * @tparam A input type of the model
+    * @tparam K type of label or class
+    * @return labels and information about the labels.
+    */
+  protected[multilabel] def labelsForPrediction[A, K](
+      example: A,
+      labelsOfInterest: GenAggFunc[A, sci.IndexedSeq[K]],
+      labelToInd: Map[K, Int]
+  ): LabelsAndInfo[K] = {
+
+    val labelsShouldPredict = labelsOfInterest(example)
+
+    val unsorted =
+      for {
+        label <- labelsShouldPredict
+        ind   <- labelToInd.get(label).toList
+      } yield (ind, label)
+
+    val problems =
+      if (labelsShouldPredict.nonEmpty) None
```

Comment:
Is this a good check? Should we be more strict about when to report errors?

Reply:
I think this seems sufficient.

Reply:
OK